### PR TITLE
Federation API for Space summary

### PR DESCRIPTION
**Pull request by richvdh**
Builds on the work done in #9643 to add a federation API for space summaries.

There's a bit of refactoring of the existing client-server code first, to avoid too much duplication.

### CODE REVIEW

File: synapse/federation/transport/server.py [hunk 1]
```diff
@@ -1474,18 +1512,24 @@ async def on_GET(self, origin, content, query, room_id):
 )
 
 
-def register_servlets(hs, resource, authenticator, ratelimiter, servlet_groups=None):
+def register_servlets(
+    hs: "synapse.server.HomeServer",
```
Comment (clokep) [inline synapse/federation/transport/server.py:1516]: Out of curiosity I always use `from synapse.app.homeserver import HomeServer` for these type hints. Should we be using this instead? (Also any reason for using the full dotted module name here instead of a `from` import? We don't seem to do that often.)

File: synapse/federation/transport/server.py [hunk 2]
```diff
@@ -1376,6 +1380,40 @@ async def on_PUT(self, origin, content, query, group_id):
...
+        self,
+        origin: str,
+        content: JsonDict,
+        query: Mapping[bytes, Sequence[bytes]],
+        room_id: str,
+    ) -> Tuple[int, JsonDict]:
+        suggested_only = content.get("suggested_only", False)
+        if not isinstance(suggested_only, bool):
+            raise SynapseError(
+                400, "'suggested_only' must be a boolean", Codes.BAD_JSON
```
Comment (clokep) [inline synapse/federation/transport/server.py:1397]: Should this (and below) be `INVALID_PARAM` instead of `BAD_JSON`?

File: synapse/handlers/space_summary.py [hunk 3]
```diff
@@ -144,12 +200,25 @@ async def _summarize_local_room(
...
+
+        # otherwise, check if the room is peekable
+        hist_vis = ""
+        hist_vis_ev = await self._state_handler.get_current_state(
+            room_id, EventTypes.RoomHistoryVisibility, ""
+        )
+        if hist_vis_ev:
+            hist_vis = hist_vis_ev.content.get("history_visibility")
+        if hist_vis == HistoryVisibility.WORLD_READABLE:
             return True
```
Comment (clokep) [inline synapse/handlers/space_summary.py]: I think it would be clearer to indent this since it can only be reached if the event exists.

File: synapse/handlers/space_summary.py [hunk 4]
```diff
@@ -76,67 +78,154 @@ async def get_space_summary(
...
+        events_result = []  # type: List[JsonDict]
+        for edge_event in itertools.islice(child_events, max_children):
+            events_result.append(
+                await self._event_serializer.serialize_event(
+                    edge_event,
+                    time_now=now,
+                    event_format=format_event_for_client_v2,
                 )
+            )
+        return (room_entry,), events_result
```
Comment (clokep) [inline synapse/handlers/space_summary.py:201]: Why wrap `room_entry` in a tuple first? Will this eventually return multiple rooms or something?

File: synapse/handlers/space_summary.py [hunk 5]
```diff
@@ -76,67 +78,154 @@ async def get_space_summary(
...
+
+        return {"rooms": rooms_result, "events": events_result}
+
+    async def federation_space_summary(
+        self,
+        room_id: str,
+        suggested_only: bool,
+        max_rooms_per_space: Optional[int],
+        exclude_rooms: Iterable[str],
+    ) -> JsonDict:
```
Comment (clokep) [inline synapse/handlers/space_summary.py:119]: Looks like this is pretty much identical to the C-S implementation, but it:

* Doesn't take a requester (and thus doesn't check if the user is in the room, falling back to whether the room is world-readable instead).
* Doesn't use `_RoomQueueEntry`.
* Falls back to the default for the maximum number of rooms per entry.

Am I missing anything else?

File: synapse/handlers/space_summary.py [hunk 6]
```diff
@@ -191,6 +280,11 @@ async def _get_child_events(self, room_id: str) -> Iterable[EventBase]:
         return (e for e in events if e.content.get("via"))
 
 
+@attr.s(frozen=True, slots=True)
+class _RoomQueueEntry:
+    room_id = attr.ib(type=str)
```
Comment (clokep) [inline synapse/handlers/space_summary.py:285]: Will the via stuff be added in a separate PR? It seems premature to separate this out.

File: synapse/federation/transport/server.py (hunk 1 above)
Comment (richvdh) [reply synapse/federation/transport/server.py:1516]: > Out of curiosity I always use from synapse.app.homeserver import HomeServer

There isn't a `HomeServer` in `synapse.app.homeserver` (except by import from `synapse.server`. Do you mean `SynapseHomeServer`? In which case, that won't really be accurate for a worker process.

> (Also any reason for using the full dotted module name here instead of a from import? We don't seem to do that often.)

Fair; I'll change this.

File: synapse/federation/transport/server.py (hunk 2 above)
Comment (richvdh) [reply synapse/federation/transport/server.py:1397]: I *think* `M_INVALID_PARAM` is for query-string params, though this is very unclear, and I've opened https://github.com/matrix-org/matrix-doc/issues/3070 about it.

File: synapse/handlers/space_summary.py (hunk 3 above)
Comment (richvdh) [reply synapse/handlers/space_summary.py]: true.

File: synapse/handlers/space_summary.py (hunk 4 above)
Comment (richvdh) [reply synapse/handlers/space_summary.py:201]: yeahh, it's mostly for symmetry with a `_summarize_remote_room` method in #9653, which can return multiple rooms.

It's worth noting that this method can also return zero rooms. I fiddled with making the whole result Optional, but it didn't really simplify anything.

File: synapse/handlers/space_summary.py (hunk 5 above)
Comment (richvdh) [reply synapse/handlers/space_summary.py:119]: That sounds about right so far. The differences get a bit bigger in #9653: it also doesn't ever call out over federation, which means it doesn't need to track `via`s (which is why it doesn't use `_RoomQueueEntry`.

File: synapse/handlers/space_summary.py (hunk 6 above)
Comment (richvdh) [reply synapse/handlers/space_summary.py:285]: sorry, yes.

File: synapse/federation/transport/server.py (hunk 1 above)
Comment (clokep) [reply synapse/federation/transport/server.py]: I suspect I've made a bunch of imports in correct! Oops. I'll put up a PR to fix this.

File: synapse/federation/transport/server.py (hunk 2 above)
Comment (clokep) [reply synapse/federation/transport/server.py:1394]: I think we're inconsistent with this in Synapse then!

File: synapse/handlers/space_summary.py (hunk 5 above)
Comment (clokep) [reply synapse/handlers/space_summary.py:119]: Got it. Thanks! I was wondering why it was separate, but that makes sense that #9653 would have this differ more.

**Review by clokep — APPROVED** [review]
Seems good!

File: synapse/federation/transport/server.py (hunk 2 above)
Comment (richvdh) [reply synapse/federation/transport/server.py:1394]: s/synapse/the entire matrix ecosystem. yes.
